tensorflow/lite/kernels/internal/optimized/optimized_ops.h

```diff
@@ -5304,56 +5304,65 @@ inline void PadImageStyle(const tflite::PadParams& op_params,
 }
 
 template <typename T>
-inline void Slice(const tflite::SliceParams& op_params,
+inline void Slice(tflite::SliceParams& op_params,
                   const RuntimeShape& input_shape,
                   const RuntimeShape& output_shape,
                   SequentialTensorWriter<T>* writer) {
   ruy::profiler::ScopeLabel label("Slice");
-  const RuntimeShape ext_shape = RuntimeShape::ExtendedShape(4, input_shape);
-  // TODO(dkalenichenko): This op only supports 4D tensors or smaller.
-  TFLITE_DCHECK_LE(op_params.begin_count, 4);
-  TFLITE_DCHECK_LE(op_params.size_count, 4);
-  const int begin_count = op_params.begin_count;
-  const int size_count = op_params.size_count;
-  // We front-pad the begin and size vectors.
-  const int start_b = 4 - begin_count > 0 ? 0 : op_params.begin[0];
-  const int stop_b = (4 - size_count > 0 || op_params.size[0] == -1)
-                         ? ext_shape.Dims(0)
-                         : start_b + op_params.size[0];
-  const int start_h = begin_count < 3 ? 0 : op_params.begin[begin_count - 3];
-  const int stop_h = (size_count < 3 || op_params.size[size_count - 3] == -1)
-                         ? ext_shape.Dims(1)
-                         : start_h + op_params.size[size_count - 3];
-  const int start_w = begin_count < 2 ? 0 : op_params.begin[begin_count - 2];
-  const int stop_w = (size_count < 2 || op_params.size[size_count - 2] == -1)
-                         ? ext_shape.Dims(2)
-                         : start_w + op_params.size[size_count - 2];
-  const int start_d = begin_count < 1 ? 0 : op_params.begin[begin_count - 1];
-  const int stop_d = (size_count < 1 || op_params.size[size_count - 1] == -1)
-                         ? ext_shape.Dims(3)
-                         : start_d + op_params.size[size_count - 1];
-
-  for (int in_b = start_b; in_b < stop_b; ++in_b) {
-    for (int in_h = start_h; in_h < stop_h; ++in_h) {
-      for (int in_w = start_w; in_w < stop_w; ++in_w) {
-        const int len = stop_d - start_d;
-        if (len > 0)
-          writer->WriteN(Offset(ext_shape, in_b, in_h, in_w, start_d), len);
-      }
-    }
-  }
+  const int slice_dimensions = input_shape.DimensionsCount();
+
+  // Below values are under condition if the input is 1-D, as it is a special
+  // case, where we will not have next axis to fetch the copy length which is
+  // input for memcpy operation below, so we are alligning the variables to the
+  // required in the case where input is 1-D. NOTE: This is done only to attain
+  // code reuse of compute_slice() defined below.
+  bool is_input_1d = slice_dimensions == 1;
```

Comment:
These logic make the code complicated. I think you can remove them by just removing the for loop in the case axis == final_axis.

Reply:
My suggested raw implementation like:
  int final_axis = input_shape.DimensionsCount() - 1;
  int copy_len = op_params.size[final_axis] - op_params.begin[final_axis];

  // Store as std::function to allow recursion.
  std::function<void(int, int)> compute_slice =
      [&compute_slice, final_axis, copy_len, writer,
       input_shape, op_params](int axis, int curr_indice) {
        if (axis == (final_axis)) {
            if (copy_len > 0) writer->WriteN(curr_indice + op_params.begin[final_axis], copy_len);
        } else {
          for (int in_x = op_params.begin[axis]; in_x < op_params.size[axis];
               in_x++) {
            int indices = (in_x + curr_indice) * input_shape.Dims(axis + 1);
            compute_slice(axis + 1, indices);
          }
        }
      };

which can be shared for both reference and optimized_ops. However, I haven't debug this. Could you take time to implement this?